Rewrite of AWS Chunked support

Randgalt opened the pull request:
Rewrite of `AwsChunkedInputStream` for simplicity/clarity
    
~~Added `ReadAheadInputStream` which buffers a portion of a delegate input stream which is assumed to be an `AwsChunkedInputStream`. The purpose of this is to catch an edge case with the final chunk. By the time the final chunk
signature is verified, the bytes of the chunk will already be sent to the remote server and, thus, it's too late to prevent it. By reading ahead, the final chunk will get validated before the bytes are sent to the remote.~~

When the end of a chunk is reached it reads the next chunk's header. This ensures we don't send the final bytes of a request before they've been validated against the signature. i.e. the final chunk's signature cannot be validated
until the final chunk has been read. But, without this change, the final chunk's bytes will already have been sent.

vagaerg commented on trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/rest/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,128 @@
+        extends InputStream
+{
+    private static final int FINAL_CHUNK_SIZE = "0;chunk-signature=%s"
+            .formatted("0".repeat(AwsS3V4ChunkSigner.getSignatureLength()))
```
Nitpick:
```java
"0;chunk-signature=".length() + AwsS3V4ChunkSigner.getSignatureLength()
```

?

vagaerg commented on trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java:
```diff
@@ -0,0 +1,332 @@
+    }
+
+    @Test
+    public void testGoodDuplicateSession()
```
Nit: `testRecreateSessionValidatesGoodPayload`?

vagaerg commented:
As an update of what we've discussed over Slack: we need to make sure we always read in a way that's not aligned to the chunk boundaries.

We need to "complete" each chunk and inspect the next one so that we never send a payload of the correct size unless we have verified all signatures:

- S3 will reject a write unless the `Content-Length` is correct (i.e., we have sent the same number of bytes we said we would)
- Our code will refuse to continue streaming if a chunk fails signature verification: this is only known after we have read the entire payload for said chunk, and gets verified again upon reading the next chunk

We therefore need to make sure that:
- We do not stream the last byte of any one chunk if its signature does not match
- We read the header of the next chunk before streaming the last byte of the previous chunk, so that we can properly validate the signature of all chunks - including the last one
    - If the next chunk is of size zero (i.e., it is the last one), we should validate its signature immediately

I do not think we can compute the number of bytes to "preload" in our input stream, as even if we were to make it a prime, the offset by which we would be misaligned with the underlying chunk size would move with each read, eventually aligning at some point.

vagaerg commented:
https://github.com/trinodb/aws-proxy/pull/103/commits/d94bc37ee8c1f94db27575335e34b2ca14d09dac

This just reads the header of the next chunk after a read if we reached its boundary.

I think it may be enough:
- It would validate the signature of the current chunk immediately after a `read` call
- `nextChunk` (at least as of that PR, prior to the nice cleanup you added here) will also immediately validate a chunk of size 0

I uncommented the tests that originally failed and they pass now. Wdyt?

Randgalt commented:
> [d94bc37](https://github.com/trinodb/aws-proxy/commit/d94bc37ee8c1f94db27575335e34b2ca14d09dac)
> 
> This just reads the header of the next chunk after a read if we reached its boundary.
> 
> I think it may be enough:

This is good and solves the problem. I've updated this PR and removed the Read head shim and altered the re-written `AwsChunkedInputStream` to read the next chunk header. I tested against `testAwsChunkedUpload()` and it fixes it.

We can either go with the previous, Apache Commons based implementation. Or use this one. wdyt?

vagaerg commented on trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/rest/AwsChunkedInputStream.java:
```diff
@@ -15,241 +15,213 @@
+            int i = delegate.read();
+            if (i < 0) {
+                delegateIsDone = true;
+                throw new EOFException("Unexpected end of stream");
```
Should we only throw here if `state != State.LAST_CHUNK`?

Randgalt replied:
this code should never run if state is `State.LAST_CHUNK`. So, it's best to leave the exception.

vagaerg commented on trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java:
```diff
@@ -0,0 +1,337 @@
+{
+    private static final Credential GOOD_CREDENTIAL = new Credential("TEST_ACCESS_KEY", "TEST_SECRET_KEY");
+    private static final String GOOD_CONTENT = "The quick brown fox jumps over the lazy dog's head";
+    private static final String GOOD_SEED = "10e300fbbad6946a482aaa7de0bdc8f592d4c372306dd746a22d18b7b66b4527";
```
Nit: this seed can be anything, as long as it isn't the same as `BAD_SEED`. Unlike in other tests, this one isn't validated at all (the seed is the signature of the request itself, which we don't check here).

What about we do `GOOD_SEED = "GOOD_DUMMY_VALUE"` and `BAD_SEED = "BAD_DUMMY_VALUE"`?

I think it makes it easier to understand what exactly `AwsChunkedInputStream` will validate and what it will not validate. GIving it a seemingly "real" seed could be a bit confusing IMO

vagaerg commented on trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java:
```diff
@@ -0,0 +1,337 @@
+    public void testChunkedInputStreamSmallBuffer()
+            throws IOException
+    {
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(CHUNKED_INPUT.getBytes(StandardCharsets.US_ASCII));
```
Nit - should we use `UTF_8` here just like in the test above?

Randgalt replied:
Yeah. I meant to get all of those.

vagaerg commented on trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java:
```diff
@@ -0,0 +1,337 @@
+            out.write(buffer, 0, len);
+        }
+        assertEquals(-1, in.read(buffer));
+        assertEquals(-1, in.read(buffer));
```
Can we also check the contents of what we've read? We do it in the test above, would be good to make sure the final contents are also correct even if our reads are unaligned with the chunk boundaries

vagaerg commented on trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java:
```diff
@@ -0,0 +1,337 @@
+        in.close();
+    }
+
+    // Missing \r\n at the end of the first chunk
```
I think this comment is wrong - this test case is actually checking 2 things right?
- The first chunk has 6 bytes, yet the header says it's 5
- The second chunk has no header

Randgalt replied:
I copied that. I've removed the comment. The test can speak for itself tbh

vagaerg commented on trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java:
```diff
@@ -0,0 +1,337 @@
+    public void testCorruptChunkedInputStreamTruncatedCRLF()
+            throws IOException
+    {
+        String s = "5;chunk-signature=0\r\n01234";
```
This is a good test, but I'd suggest checking:
- `5;chunk-signature=0\r\n01234` (the current one): valid chunk without either a CRLF nor a final chunk
- `5;chunk-signature=0\r\n01234\r\n`: valid chunk without a final chunk
- `5;chunk-signature=0\r\n012340;chunk-signature=0\r\n\r\n`: valid chunk without a CRLF between it and the next (Final) chunk

vagaerg commented:
> We can either go with the previous, Apache Commons based implementation or use this one. wdyt?

I think this current implementation is cleaner and more readable, happy to go with it

vagaerg commented on trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java:
```diff
@@ -0,0 +1,348 @@
+            throws IOException
+    {
+        // altered to add a few more bad stings
+        Stream.of("5;chunk-signature=0\r\n01234", ";chunk-signature=0\\r\\n01234\\r\\n", "5;chunk-signature=0\\r\\n012340;chunk-signature=0\\r\\n\\r\\n")
```
The `\` in `\r` and `\n` should not be escaped in any case

Randgalt replied:
Oops - IntelliJ did that.

vagaerg commented on trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java:
```diff
@@ -0,0 +1,348 @@
+    public void testHugeChunk()
+            throws IOException
+    {
+        ByteArrayInputStream inputStream = new ByteArrayInputStream("499602D2;chunk-signature=0\r\n01234567".getBytes(UTF_8));
```
This is the one I'm conflicted about - we check that:
- All streams must have a valid final chunk
- All chunk signatures are valid prior to streaming the last byte in the chunk

We have a test to verify that if a chunk contains more data than its header indicates, we'll throw.

However, if the header indicates the chunks are larger than they actually are we are potentially allowing unsigned content to go through.

I'm wondering whether we should have `AwsChunkedInputStream` check the `X-Amz-Decoded-Content-Length` and force a validation prior to reading the last byte. Otherwise we could be in a situation where someone can send in a request with `aws-chunked` encoding and a much larger chunk size than `X-Amz-Decoded-Content-Length` - we will never validate the signature of the chunk

Randgalt replied:
Your concern is processing more bytes than `X-Amz-Decoded-Content-Length` says we should?

vagaerg replied:
My concern is having a single chunk with a header stating a much larger chunk size than X-Amz-Decoded-Content-Length and some invalid signature

this chunk can contain X-Amz-Decoded-Content-Length bytes (which as stated would be much smaller than the chunk size stated in the chunk header)

we would stream all bytes to S3 until we get an EOF, but by this point S3 has processed the payload already without any signature verification

Randgalt replied:
it would be the same with 2 chunks. Chunk one has a reasonable size and the second a huge size. I think if we have `AwsChunkedInputStream` take an optional max bytes argument we can prevent this.

Randgalt replied:
However, this means we need to track `X-Amz-Decoded-Content-Length` (probably in `RequestContent`). I suggest that we do this as a follow-up PR. OK?

Randgalt replied:
https://github.com/trinodb/aws-proxy/issues/106
